packages/react/src/QuestionnaireForm/QuestionnaireForm.tsx

```diff
@@ -656,3 +748,95 @@ function addTargetTypes(item: QuestionnaireItem): string[] {
   const targets = extensions.map((e) => e.valueCodeableConcept?.coding?.[0]?.code) as string[];
   return targets;
 }
+
+function repeatableLinkId(linkId: string): string {
+  const repeatablePattern = /-repeatable-(\d+)$/;
+  const match = repeatablePattern.exec(linkId);
+
+  if (match) {
+    const currentRepeatableIndex = parseInt(match[1], 10);
+    return linkId.replace(repeatablePattern, `-repeatable-${currentRepeatableIndex + 1}`);
+  } else {
+    return `${linkId}-repeatable-1`;
+  }
+}
+
+function allowRepeatable(item: QuestionnaireItem, answers: Record<string, QuestionnaireResponseItemAnswer>): boolean {
+  if (item.type === QuestionnaireItemType.group) {
+    return true;
+  }
+  const linkId = item.linkId;
+  if (!linkId) {
+    return false;
+  }
+  return !!answers[linkId];
+}
+
+function repeatableInsert(
+  items: QuestionnaireItem[],
+  currentItem: QuestionnaireItem,
+  newItem: QuestionnaireItem
+): QuestionnaireItem[] {
+  for (let i = 0; i < items.length; i++) {
+    if (items[i].linkId === currentItem.linkId) {
+      const updatedItems = [...items.slice(0, i + 1), newItem, ...items.slice(i + 1)];
+      return updatedItems;
+    }
+
+    if (items[i].item) {
+      const updatedNestedItems = repeatableInsert(items[i].item ?? [], currentItem, newItem);
+      if (updatedNestedItems !== items[i].item) {
+        items[i].item = updatedNestedItems;
+      }
+    }
+  }
+  return items;
+}
+
+function removeRecentItem(items: QuestionnaireItem[], currentItem: QuestionnaireItem): QuestionnaireItem[] {
+  for (let i = 0; i < items.length; i++) {
+    if (items[i].linkId === currentItem.linkId) {
+      items.splice(i, 1);
+
+      if (i - 1 >= 0) {
+        items[i - 1].repeats = true;
+      }
+      return items;
+    }
+
+    if (items[i].item) {
+      const updatedNestedItems = removeRecentItem(items[i].item ?? [], currentItem);
+      if (updatedNestedItems !== items[i].item) {
+        items[i].item = updatedNestedItems;
+      }
+    }
+  }
+  return items;
+}
+
+function createRepeatableItem(item: QuestionnaireItem): QuestionnaireItem {
+  let newText = item.text ?? '';
+  if (!newText.endsWith(' continued')) {
+    newText += ' continued';
+  }
+  return {
+    ...item,
+    text: newText,
+    linkId: repeatableLinkId(item.linkId ?? ''),
```

Comment:
I'm pretty sure we should not be modifying `linkId` inside of `QuestionnaireForm`.  `linkId` has a pretty well defined contract in FHIR Questionnaires.

Reply:
See this discussion for `linkId` behavior with `repeats`: http://community.fhir.org/t/questionnaire-repeating-groups-what-is-the-correct-format/2276

Reply:
"Repeating groups appear multiple times. Repeating questions have multiple answers"

Reply:
https://hl7.org/fhir/R4/questionnaireresponse.html